### PR TITLE
Issue 23078 unable to add categories to bundles

### LINKED ISSUE #23078: Unable to add categories to bundles
We are not allowed to add categories to a bundle in order to push. When you try this, we are not getting errors but the category is not added to the existing bundle. 

FYI: No errors ithe n the log, neither in console 

- Login as an admin 
- Configure your push environment 
- Go to Content Model -- > Categories 
- Select any category and try to add to a bundle 
- Check the bundle and make sure the category was added 

### Expected behavior
We need to be able to add categories to bundles and push without problems 

### Desktop (please complete the following information):
Tested on release-22.10 // Docker // FF 

### Acceptance criteria
- We need to be able to add categories to bundles
- Push Categories

### PR DESCRIPTION
**Pull request by freddyucv**
### Proposed Changes
* When we add a category into the Bundle before this change we added the 'CAT' word into the bundle and later when the bundle is genereted it added all the Top Categories into the bundle
* Now when we add the Category the Category's inode is added and later all the other top categories are added as dependencies.

https://github.com/dotCMS/core/compare/issue-23078-Unable-to-add-categories-to-bundles?expand=1#diff-45822bf92f039131f8acdcd3294b73b410249dec48c45cba2bd88cceff6a116eL1014-R1060

* When a Category is added directly by the user into the bundle then We add it into the manifest with right reason

https://github.com/dotCMS/core/compare/issue-23078-Unable-to-add-categories-to-bundles?expand=1#diff-292aa9b2b6ead8a5061b3ad0b4465cced09f4ba4771217cec2d5845a172e1ddfR314

* When the others categories are added, they are added as Dependency into the Manifest

https://github.com/dotCMS/enterprise/pull/890/files

### CODE REVIEW

**Comment by wezell** [comment]
@freddyucv  My understanding was that when we pushed categories, we pushed all categories. This is because a child category is dependent on its parents.  Are we still doing this?  Or does this recurse up the category tree adding the parent categories of the selected categories?  Or is it just pushing the the top level (which would be wrong)

**Comment by freddyucv** [comment]
> @freddyucv My understanding was that when we pushed categories, we pushed all categories. This is because a child category is dependent on its parents. Are we still doing this? Or does this recurse up the category tree adding the parent categories of the selected categories? Or is it just pushing the the top level (which would be wrong)

Yes, we are doing the same, the different now is that we are saving the category what we added into the publishing_queue table so now we are not getting a empty bundle when we list it into the publishing queque portlet.

But later we added all the Categories when the bundle is build.

The other different now is that we are registering all in the manifest file, so the category add directly is mark as "Add directly by the user", and the others are mark as "Dependency from: [The one add by the user]", and the children are mark as dependency from its parent

**Comment by dsilvam** [comment]
We need to change the option "Add to Bundle" to something generic like "Add All To Bundle", and then in the bundle we should not show any cat in particular but a general text like "Syncing All Categories"

File: dotCMS/src/main/webapp/WEB-INF/messages/Language.properties [hunk 1]
```diff
@@ -117,6 +117,7 @@ add-tag = Add Tag
 Add-Template = Add Template
 add-template = Add Template
 Add-To-Bundle = Add To Bundle
+Add-All-Categories-To-Bundle = Add all categories to bundle
```
Comment (dsilvam) [inline dotCMS/src/main/webapp/WEB-INF/messages/Language.properties:120]: Capital C and B. "Add all Categories to Bundle"

Comment (freddyucv) [reply dotCMS/src/main/webapp/WEB-INF/messages/Language.properties]: done https://github.com/dotCMS/core/pull/23183/commits/f391675dc6195856bc48cfd178fd3253b1177121

File: dotCMS/src/integration-test/java/com/dotcms/enterprise/publishing/remote/bundler/FileBundlerTestUtil.java [hunk 2]
```diff
@@ -117,14 +120,22 @@ public static File getFolderFilePath(final Folder folder, final File bundleRoot)
     public static File getCategoryPath(final Category category, final File bundleRoot)
             throws DotSecurityException, DotDataException {
 
+        final String bundleId = bundleRoot.getAbsolutePath()
+                .substring(bundleRoot.getPath().lastIndexOf("/") + 1);
```
Comment (jdotcms) [inline dotCMS/src/integration-test/java/com/dotcms/enterprise/publishing/remote/bundler/FileBundlerTestUtil.java:124]: Use StringPool.FORWARD_SLASH

File: dotCMS/src/integration-test/java/com/dotcms/enterprise/publishing/remote/bundler/FileBundlerTestUtil.java [hunk 3]
```diff
@@ -117,14 +120,22 @@ public static File getFolderFilePath(final Folder folder, final File bundleRoot)
...
+
         final String liveWorking = hasLiveVersion(category.getIdentifier()) ? "live" : "working";
 
         final String categoryFilePath = bundleRoot.getPath()
                 + File.separator
                 + liveWorking + File.separator
                 + APILocator.getHostAPI().findSystemHost().getHostname()
                 + File.separator
-                + category.getInode() + ".category.dpc.xml";
+                + category.getInode() + (categoryAddDirectly ? ".category.xml" : ".category.dpc.xml");
```
Comment (jdotcms) [inline dotCMS/src/integration-test/java/com/dotcms/enterprise/publishing/remote/bundler/FileBundlerTestUtil.java:138]: these should be constants

File: dotCMS/src/integration-test/java/com/dotcms/publishing/manifest/CSVManifestBuilderTest.java [hunk 4]
```diff
@@ -181,7 +181,7 @@ private static TestCase getFolderTestCase() {
     private static TestCase getCategoryTestCase() {
         final Category category = new CategoryDataGen().nextPersisted();
         final String line = list(PusheableAsset.CATEGORY.getType(), category.getInode(), StringPool.BLANK,
-                 category.getTitle(), "", "")
+                 category.getCategoryName(), "", "")
```
Comment (jdotcms) [inline dotCMS/src/integration-test/java/com/dotcms/publishing/manifest/CSVManifestBuilderTest.java:184]: StringPool.Blank

File: dotCMS/src/main/java/com/dotcms/publisher/business/PublishQueueElementTransformer.java [hunk 5]
```diff
@@ -83,6 +84,8 @@ private static Map<String, Object> getMap(final String id, final String type,
             result = getMapForContentlet(id);
         } else if (isLanguage(type)) {
             result = getMapForLanguage(id);
+        } else if ("CAT".equals(id)) {
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/publisher/business/PublishQueueElementTransformer.java:87]: this should be a constant

File: dotCMS/src/main/java/com/dotcms/publisher/business/PublishQueueElementTransformer.java [hunk 6]
```diff
@@ -113,6 +116,10 @@ private static Map<String, Object> getMap(final String id, final String type,
         return result;
     }
 
+    private static boolean isCategory(final String type) {
+        return type.equals(PusheableAsset.CATEGORY.getType());
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/publisher/business/PublishQueueElementTransformer.java:120]: probably the constant should be at the left side such as 

PusheableAsset.CATEGORY.getType().equals(type)

File: dotCMS/src/main/java/com/dotcms/publisher/business/PublishQueueElementTransformer.java [hunk 7]
```diff
@@ -126,6 +133,10 @@ private static Map<String, Object> getMapForLanguage(final String id) {
             ) : map(TITLE_KEY, id);
     }
 
+    private static Map<String, Object> getMapForCategory(){
+        return map(TITLE_KEY, "Syncing All Categories");
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/publisher/business/PublishQueueElementTransformer.java:137]: i18n?

File: dotCMS/src/main/java/com/dotcms/publisher/pusher/PushPublisherConfig.java [hunk 8]
```diff
@@ -134,6 +134,27 @@ public boolean justIncludesUsers() {
...
+	 *
+	 * @return
+	 */
+	public boolean justIncludesCategories() {
+
+		if (!UtilMethods.isSet(this.getAssets())) {
+			return false;
+		}
+
+		for (PublishQueueElement asset : this.getAssets()) {
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/publisher/pusher/PushPublisherConfig.java:149]: set to final

Comment (jdotcms) [reply dotCMS/src/main/java/com/dotcms/publisher/pusher/PushPublisherConfig.java:149]: you may reduce this to return !this.getAssets().stream.anyMatch (asset -> !"CAT".equalsIgnoreCase(asset.getAsset()))

File: dotCMS/src/main/java/com/dotcms/publisher/util/dependencies/DependencyManager.java [hunk 9]
```diff
@@ -311,6 +313,12 @@ public void setDependencies()
 				final String userID = getUserID(asset);
 				final User user = APILocator.getUserAPI().loadUserById(userID);
 				config.add(user, PusheableAsset.USER, ManifestReason.INCLUDE_BY_USER.getMessage());
+			} else if ("CAT".equals(asset.getAsset())) {
+				config.writeIncludeManifestItem((ManifestItem) () -> new ManifestInfoBuilder()
+						.objectType(PusheableAsset.CATEGORY.getType())
+						.title("Syncing All Categories")
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/publisher/util/dependencies/DependencyManager.java:319]: i18n

**Review by jdotcms — APPROVED** [review]

File: dotCMS/src/integration-test/java/com/dotcms/enterprise/publishing/remote/bundler/FileBundlerTestUtil.java (hunk 3 above)
Comment (freddyucv) [reply dotCMS/src/integration-test/java/com/dotcms/enterprise/publishing/remote/bundler/FileBundlerTestUtil.java:138]: this is a test class so I think it is ok to be hardcoding

File: dotCMS/src/integration-test/java/com/dotcms/enterprise/publishing/remote/bundler/FileBundlerTestUtil.java (hunk 2 above)
Comment (freddyucv) [reply dotCMS/src/integration-test/java/com/dotcms/enterprise/publishing/remote/bundler/FileBundlerTestUtil.java]: done https://github.com/dotCMS/core/pull/23183/commits/efcc202067d1ebc9384341b6869f6b78aa350bc7#diff-17175f0a859646e450f675099713c3f0e1d2817928a7ebdee304463353a5ddccR125

File: dotCMS/src/integration-test/java/com/dotcms/publishing/manifest/CSVManifestBuilderTest.java (hunk 4 above)
Comment (freddyucv) [reply dotCMS/src/integration-test/java/com/dotcms/publishing/manifest/CSVManifestBuilderTest.java]: done https://github.com/dotCMS/core/pull/23183/commits/efcc202067d1ebc9384341b6869f6b78aa350bc7#diff-d2fbe7006d2de513acebc78bd7b762e626273a7ac2c7e0471217bf6785c71afcR184

File: dotCMS/src/main/java/com/dotcms/publisher/business/PublishQueueElementTransformer.java (hunk 5 above)
Comment (freddyucv) [reply dotCMS/src/main/java/com/dotcms/publisher/business/PublishQueueElementTransformer.java]: done https://github.com/dotCMS/core/pull/23183/commits/49a6bcf7c5d48d87d283555e1283b87a17e75169

File: dotCMS/src/main/java/com/dotcms/publisher/business/PublishQueueElementTransformer.java (hunk 6 above)
Comment (freddyucv) [reply dotCMS/src/main/java/com/dotcms/publisher/business/PublishQueueElementTransformer.java]: really I dont need this method anymore https://github.com/dotCMS/core/pull/23183/commits/efcc202067d1ebc9384341b6869f6b78aa350bc7#diff-3389c7193d609b29dbc719b8c559e5149211bb7b105aa9bee2eb84b1c7f733b1L119

File: dotCMS/src/main/java/com/dotcms/publisher/business/PublishQueueElementTransformer.java (hunk 7 above)
Comment (freddyucv) [reply dotCMS/src/main/java/com/dotcms/publisher/business/PublishQueueElementTransformer.java]: done https://github.com/dotCMS/core/pull/23183/commits/efcc202067d1ebc9384341b6869f6b78aa350bc7#diff-3389c7193d609b29dbc719b8c559e5149211bb7b105aa9bee2eb84b1c7f733b1R136

File: dotCMS/src/main/java/com/dotcms/publisher/pusher/PushPublisherConfig.java (hunk 8 above)
Comment (freddyucv) [reply dotCMS/src/main/java/com/dotcms/publisher/pusher/PushPublisherConfig.java]: https://github.com/dotCMS/core/pull/23183/commits/efcc202067d1ebc9384341b6869f6b78aa350bc7#diff-28ca72a88db7305591d8dac2158e7eb1c45b87e08805666f31c0f0be5b4089cfR149

File: dotCMS/src/main/java/com/dotcms/publisher/util/dependencies/DependencyManager.java (hunk 9 above)
Comment (freddyucv) [reply dotCMS/src/main/java/com/dotcms/publisher/util/dependencies/DependencyManager.java]: done https://github.com/dotCMS/core/pull/23183/commits/49a6bcf7c5d48d87d283555e1283b87a17e75169#diff-292aa9b2b6ead8a5061b3ad0b4465cced09f4ba4771217cec2d5845a172e1ddfR321

**Review by dsilvam — APPROVED** [review]

**Pull request merged**